sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -494,7 +506,27 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case DateType =>
       buildCast[Int](_, d => null)
     case TimestampType =>
-      buildCast[Long](_, t => timestampToLong(t).toShort)
+      buildCast[Long](_, t => {
+        val longValue = timestampToLong(t)
+        if (!failOnIntegerOverflow || longValue == longValue.toShort) {
+          longValue.toShort
+        } else {
+          throw new ArithmeticException(s"Casting $t to Short causes overflow.")
```

Comment:
btw, this error is a kind of overflows? Is this not an out-of-range error as pgSQL says so?
```
postgres=# select cast(float4(1e24) as bigint);
ERROR:  bigint out of range
```

Reply:
I was following the error message from https://github.com/apache/spark/pull/21599.